tests/utils/utils.go

```diff
@@ -0,0 +1,351 @@
+		for {
+			block, err := client.BlockByNumber(cctx, nil)
+			Expect(err).Should(BeNil())
+			if block.NumberU64() >= height {
+				log.Info("client accepted the block containing SendWarpMessage", "client", i, "height", block.NumberU64())
+				break
+			}
+		}
+	}
+}
+
+// Constructs the aggregate signature, packs the Teleporter message, and relays to the destination
+// Returns the receipt on the destination chain
+func RelayMessage(
+	ctx context.Context,
+	sourceBlockHash common.Hash,
+	sourceBlockNumber *big.Int,
+	source SubnetTestInfo,
+	destination SubnetTestInfo,
+) *types.Receipt {
+	log.Info("Fetching relevant warp logs from the newly produced block")
+	logs, err := source.ChainWSClient.FilterLogs(ctx, interfaces.FilterQuery{
+		BlockHash: &sourceBlockHash,
+		Addresses: []common.Address{warp.Module.Address},
+	})
+	Expect(err).Should(BeNil())
+	Expect(len(logs)).Should(Equal(1))
+
+	// Check for relevant warp log from subscription and ensure that it matches
+	// the log extracted from the last block.
+	txLog := logs[0]
+	log.Info("Parsing logData as unsigned warp message")
+	unsignedMsg, err := avalancheWarp.ParseUnsignedMessage(txLog.Data)
+	Expect(err).Should(BeNil())
+
+	// Set local variables for the duration of the test
+	unsignedWarpMessageID := unsignedMsg.ID()
+	unsignedWarpMsg := unsignedMsg
+	log.Info("Parsed unsignedWarpMsg", "unsignedWarpMessageID", unsignedWarpMessageID, "unsignedWarpMessage", unsignedWarpMsg)
+
+	// Loop over each client on chain A to ensure they all have time to accept the block.
+	// Note: if we did not confirm this here, the next stage could be racy since it assumes every node
+	// has accepted the block.
+	WaitForAllValidatorsToAcceptBlock(ctx, source.ChainNodeURIs, source.BlockchainID, sourceBlockNumber.Uint64())
+
+	// Get the aggregate signature for the Warp message
+	log.Info("Fetching aggregate signature from the source chain validators")
+	warpClient, err := warpBackend.NewClient(source.ChainNodeURIs[0], source.BlockchainID.String())
+	Expect(err).Should(BeNil())
+	signedWarpMessageBytes, err := warpClient.GetAggregateSignature(ctx, unsignedWarpMessageID, params.WarpQuorumDenominator)
+	Expect(err).Should(BeNil())
+
+	// Construct the transaction to send the Warp message to the destination chain
+	signedTx := CreateReceiveCrossChainMessageTransaction(
+		ctx,
+		signedWarpMessageBytes,
+		big.NewInt(1),
+		teleporterContractAddress,
+		fundedAddress,
+		fundedKey,
+		destination.ChainWSClient,
+		destination.ChainIDInt,
+	)
+
+	log.Info("Sending transaction to destination chain")
+	receipt := SendTransactionAndWaitForAcceptance(ctx, destination.ChainWSClient, signedTx)
+
+	bind, err := teleportermessenger.NewTeleportermessenger(teleporterContractAddress, source.ChainWSClient)
+	Expect(err).Should(BeNil())
+	// Check the transaction logs for the ReceiveCrossChainMessage event emitted by the Teleporter contract
+	for _, log := range receipt.Logs {
```

Comment:
any specific reason we iterate through these logs and then add a `Expect(false).Should(BeTrue()`, instead of similar to when we initially send message on source chain and just check parsing `receipt.Logs[0]`?

Reply:
The `ReceiveCrossChainMessage` event isn't guaranteed to be in index 0, so we iterate over all of the logs and check that one of them is the expected event. If none match, then we assert.

Reply:
I added helpers to iterate over a slice of logs, and return the send/receive logs if present. Else, returns an error.

Reply:
sorry one more question to clarify own understanding, why isn't the `ReceiveCrossChainMessage` event guaranteed to be at index 0? Shouldn't the transaction receipt from the `receiveCrossChainMessage` destination tx only have the `ReceiveCrossChainMessage` event?

Reply:
There are three logs in this case. One of them is the `ReceiveCrossChainMessage` event, another is the `FailedMessageExecution` event (we are sending a dummy payload). I'm not sure what the third is, probably something else related to failed message execution.

Reply:
ah ye you're right, thanks for explaining